### ISSUE TITLE
train loss decrease.  The model also learn a mask under single source architecture ?

### ISSUE THREAD

**Issue by mysee1989** [oawiles/FAb-Net#1]
When i cropped  voxceleb1(voxceleb2 is being prepared but not prepared) and started training a Fab-net according to your shared code. 
I observed decrease in train loss before 35 epoches. After that,  both train and validation loss saturated.

![image](https://user-images.githubusercontent.com/3388319/45144270-3ee0fe00-b1f0-11e8-8964-b9cb55e67d6c.png)
Purple curve denotes train loss. 

Curriculum learning has been training procedure.  I have tried other transform methods on input images.(enlarge rand crop size etc.)
Can you give me some other suggestions for training?  

One of my input images are as follows:
![0000875](https://user-images.githubusercontent.com/3388319/45144617-3210da00-b1f1-11e8-8194-4ce71d4b5646.jpg)

**Comment by oawiles** [comment]
We cropped quite tightly around the face. Look at the
examples in the paper for an idea of the crops (we had to take tight crops
to make them consistent between datasets --e.g. Vox1/2).

A few other points:
1.  I'd highly recommend also using Vox2 (Vox1 is smaller and has less variation in pose/expression).
When we trained on Vox1, we trained for < 1000 iterations as it converged fast.
On Vox2 it also trained quite quickly (if I remember it took about a day to train I think): we trained for about 3000 iterations but it plateaued before that.

2. Please check the batch size for curriculum -- the number we used is in the paper.

3. Our models also saturate training/validation loss. I'm not sure what you're asking but the loss won't go to zero as there are some transformations that the warp field wouldn't be able to model (even on train), as an object may appear that was not in the other frame.

Finally, I think it looks fine actually: the loss will be noisy. But it looks like the loss is still going down (albeit slowly). You can change the running average in tensorboard to get a better idea of whether it is converged, irrespective of the noise.

Please let me know if you still have questions.

Olivia

**Comment by mysee1989** [comment]
Thank you for your repley. I also think that tight and consistent crops between datasets are essential. 

According to example in your paper(FAb-Net),  It's  possible that my crops can meet the requirements. Thank you very much if you could look up the exact crops.

While i'm still confused about your transform method in VoxCelebData_withmask.py, as follows:

`crop = 170`
`precrop = crop + 20`
`self.transform = Compose([Scale((256,256)),`
                                         ` Pad((20,80,20,30)),`
                                         ` CenterCrop(precrop), RandomCrop(crop),`
                                          `Scale((256,256)), ToTensor()])`

Since original input image has been **tightly cropped**,  I  think  the Pad + CenterCrop + RandomCrop operation will cause same facial parts missing..
since the operation will transfrom img(256x256) to
256x256(input) -> 296x366(pad) -> 190x190(centercrop) -> 170x170(randomcrop) -> 256x256(scale)
 Will it influence the training?

**Comment by oawiles** [comment]
I updated my previous comment with more points: please look at it. 

As for your next question: I'd argue some jitter, etc is good for training! It makes the model more robust to these transformations in the real world. I wouldn't expect it to be too much of a problem.

The reason we have this is as I mentioned above, to make the crops on Vox1/2 to be consistent which required taking a smaller, tighter crop.

**Comment by mysee1989** [comment]
Thank you for your suggestions & explanations.  I'll  try..

**Comment by mysee1989** [comment]
After taking your advice, i have successfully trained a good model based on multi-source architecture. Thank you.

**Comment by mysee1989** [comment]
When i train a model on single-source architecute, i found that actually the model also learn a mask ?
Code  in multiview_maskfaceloss function (train_attention_curriculum.py) is as follows:
`sampled_images = torch.cat(xc_sampled_images, 4)`
	`attention_maps = torch.cat(attention_maps, 4)`
	`result_image = (sampled_images * attention_maps.exp()).sum(dim=4) / `attention_maps.exp().sum(dim=4)`

I tried to delete mask under single source architecute. performance and visualization degrades.
Can you explain why ?  Thank you.

**Comment by oawiles** [comment]
Hi,

Yes we did write the code this way in order to facilitate trying with
different views and having the same architecture to ease tasting, but it
shouldn't make a big difference as the confidence maps are normalised so
the maps would be normalised to 1 everywhere.

I can't be sure why performance has changed a bit, as I don't know what
you've changed, but I wouldn't expect a massive difference.

However if there is some difference, that is quite interesting and the
reason I would expect (without any actual experimental investigation) is
that the confidence maps act as a regulariser on the weights.

Hope that helps.

Olivia

**Comment by mysee1989** [comment]
Thank you for your explanation.  At last there is little difference between them( with or without confidence map in single-source architecture). 
 Anyway, Fab-net is a solid & novel work and easy to reproduce.
